apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        }
+
+        _staking.slash(_user, _challenger, _amount);
+    }
+
+    /**
+    * @dev Unlock and slash a number of tokens in the staking pool from a user in favour of a challenger
+    * @param _staking Staking pool for the ERC20 token to be unlocked and slashed
+    * @param _user Address of the user to be unlocked and slashed
+    * @param _unlockAmount Number of collateral tokens to be unlocked
+    * @param _challenger Address receiving the slashed tokens
+    * @param _slashAmount Number of collateral tokens to be slashed
+    */
+    function _unlockAndSlashBalance(Staking _staking, address _user, uint256 _unlockAmount, address _challenger, uint256 _slashAmount) internal {
+        if (_unlockAmount != 0 && _slashAmount != 0) {
+            _staking.unlockAndSlash(_user, _unlockAmount, _challenger, _slashAmount);
+        } else {
+            _unlockBalance(_staking, _user, _unlockAmount);
+            _slashBalance(_staking, _user, _challenger, _slashAmount);
+        }
+    }
+
+    /**
+    * @dev Transfer tokens to an address
+    * @param _token ERC20 token to be transferred
+    * @param _to Address receiving the tokens
+    * @param _amount Number of tokens to be transferred
+    */
+    function _transfer(ERC20 _token, address _to, uint256 _amount) internal {
+        if (_amount > 0) {
+            require(_token.safeTransfer(_to, _amount), ERROR_TOKEN_TRANSFER_FAILED);
+        }
+    }
+
+    /**
+    * @dev Transfer tokens from an address to this Agreement
+    * @param _token ERC20 token to be transferred
+    * @param _from Address transferring the tokens
+    * @param _amount Number of tokens to be transferred
+    */
+    function _transferFrom(ERC20 _token, address _from, uint256 _amount) internal {
+        if (_amount > 0) {
+            require(_token.safeTransferFrom(_from, address(this), _amount), ERROR_TOKEN_DEPOSIT_FAILED);
+        }
+    }
+
+    /**
+    * @dev Approve arbitration fee tokens to an address
+    * @param _token ERC20 token used for the arbitration fees
+    * @param _to Address to be approved
+    * @param _amount Number of `_arbitrationFeeToken` tokens to be approved
+    */
+    function _approveArbitratorFeeTokens(ERC20 _token, address _to, uint256 _amount) internal {
+        require(_token.safeApprove(_to, _amount), ERROR_TOKEN_APPROVAL_FAILED);
+    }
+
+    /**
+    * @dev Change Agreement settings
+    * @param _title String indicating a short description
+    * @param _content Link to a human-readable text that describes the initial rules for the Agreements instance
+    * @param _arbitrator Address of the IArbitrator that will be used to resolve disputes
+    */
+    function _newSetting(string _title, bytes _content, IArbitrator _arbitrator) internal {
+        require(isContract(address(_arbitrator)), ERROR_ARBITRATOR_NOT_CONTRACT);
+
+        uint256 id = settingsLength++;
+        settings[id] = Setting({ title: _title, content: _content, arbitrator: _arbitrator });
```

Comment:
Might be worth doing a quick test; I think setting them separately:

```solidity
Setting storage setting = settings[id];
setting.title = ...
```

Is a bit cheaper (avoids a memory alloc when doing it like this).